maxq10xx/wolfssl-lwip-client.c

```diff
@@ -0,0 +1,318 @@
+{
+    (void)ssl;
+    (void)hint;
+    (void)key;
+    (void)key_max_len;
+
+#ifdef WOLFSSL_PSK_MULTI_ID_PER_CS
+    /* Multiple calls for each cipher suite. First identity byte indicates the
+     * number of identites seen so far for cipher suite. */
+    if (identity[0] != 0) {
+        return 0;
+    }
+#endif  /* WOLFSSL_PSK_MULTI_ID_PER_CS */
+
+    /* see internal.h MAX_PSK_ID_LEN for PSK identity limit */
+    XSTRNCPY(identity, kIdentityStr, id_max_len);
+    XSTRNCAT(identity, ciphersuite + XSTRLEN(ciphersuite) - 6, id_max_len);
+
+    /* For TLS 1.3, we just return an unmodified key. */
+    return 32;
+}
+#endif /* WOLFSSL_TLS13 */
+
+void mxc_wolfssl_create(int sockfd, WOLFSSL **ssl)
+{
+    int ret;
+    int exit_clean = 0;
+
+    /* declare wolfSSL objects */
+    WOLFSSL_CTX* ctx = NULL;
+
+    /*---------------------------------------------------*/
+    /* Start of wolfSSL initialization and configuration */
+    /*---------------------------------------------------*/
+#ifdef DEBUG_WOLFSSL
+    wolfSSL_Debugging_ON();
+#endif
+
+    /* Initialize wolfSSL */
+    if ((ret = wolfSSL_Init()) != WOLFSSL_SUCCESS) {
+        MXC_ERR_MSG("ERROR: Failed to initialize the library\n");
+        exit_clean = 1;
+        goto exit;
+    }
+
+    /* Create and initialize WOLFSSL_CTX */
+#ifdef WOLFSSL_TLS13
+    if (tls_version == TLS13) {
+        MXC_DEBUG_MSG_GRN("TLS v1.3\n");
+        ctx = wolfSSL_CTX_new(wolfTLSv1_3_client_method());
+    } else
+#endif /* WOLFSSL_TLS13 */
+    {
+        MXC_DEBUG_MSG_GRN("TLS v1.2\n");
+        ctx = wolfSSL_CTX_new(wolfTLSv1_2_client_method());
+    }
+
+    if (ctx == NULL) {
+        MXC_ERR_MSG("ERROR: failed to create WOLFSSL_CTX\n");
+        ret = -1;
+        exit_clean = 1;
+        goto exit;
+    }
+
+    /* At this point you would normally register a CA certificate, however, it
+     * resides in MAXQ10xx and has already been registered. */
+#ifndef NO_FILESYSTEM
+    /* Load the dummy private key; actually a public key. The actual private
+     * key resides in MAXQ 10xx. */
+    if (key_file != NULL) {
+        if ((ret = wolfSSL_CTX_use_PrivateKey_file(ctx, key_file,
+                            WOLFSSL_FILETYPE_PEM)) != WOLFSSL_SUCCESS) {
+            MXC_ERR_MSG("ERROR: failed to load %s, please check the "
+                    "file.\n", key_file);
+            exit_clean = 1;
+            goto exit;
+        }
+    }
+#else
+
+#if defined (TEST_RSA)
+ret = wolfSSL_CTX_use_PrivateKey_buffer(ctx, rsa_key_der_2048,
+            sizeof(rsa_key_der_2048), WOLFSSL_FILETYPE_ASN1);
+#elif defined (HAVE_ECC)
+    ret = wolfSSL_CTX_use_PrivateKey_buffer(ctx, ecc_key_der_256,
+            sizeof(ecc_key_der_256), WOLFSSL_FILETYPE_ASN1);
+#else
+    ret = wolfSSL_CTX_use_PrivateKey_buffer(srv_ctx, server_key_der_2048,
+            sizeof_server_key_der_2048, WOLFSSL_FILETYPE_ASN1);
+#endif
+
+    if (ret != WOLFSSL_SUCCESS) {
+        MXC_DEBUG_MSG("error loading private key\n");

```

Comment:
why no `exit_clean = 1` in this error case? It also seems like all errors set exit_clean, maybe consider having exit_clean default to 1 and only one case of setting exit_clean to 0 if it reaches successful state just before the goto label.